File: src/integration-test/java/org/radarcns/redcap/integration/IntegratorTest.java
package org.radarcns.redcap.integration;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mockito;
import org.radarcns.exception.TokenException;
import org.radarcns.redcap.config.RedCapInfo;
import org.radarcns.redcap.config.RedCapManager;
import org.radarcns.redcap.managementportal.Subject;
import org.radarcns.redcap.util.RedCapClient;
import org.radarcns.redcap.util.RedCapTrigger;

import java.io.IOException;
import java.net.URISyntaxException;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static org.junit.Assert.assertEquals;
import static org.radarcns.redcap.util.IntegrationUtils.*;

public class IntegratorTest {
    private Map<String, String> testAttributes = new HashMap<>();
    private RedCapClient redCapClient;
    private RedCapIntegator redCapIntegrator;
    private MpIntegrator mpIntegrator = new MpIntegrator(mpClient);
    private List<String> attributeKeys;
    public static RedCapTrigger trigger = new RedCapTrigger(TRIGGER_BODY);
    public static final RedCapInfo redCapInfo = RedCapManager.getInfo(trigger);




    public static final String REDCAP_ATTRIBUTE_1 = "treatment_arm";
    public static final String REDCAP_ATTRIBUTE_1_VAL = "arm1";
    public static final String REDCAP_ATTRIBUTE_1_VAL_2 = "arm2";
    public static final String REDCAP_ATTRIBUTE_2 = "another_attribute";
    public static final String REDCAP_ATTRIBUTE_2_VAL = "test";

    public static final String HUMAN_READABLE_ID_KEY = "Human-readable-identifier";
    public static final String HUMAN_READABLE_ID = WORK_PACKAGE + "-" + MP_PROJECT_ID + "-" + MP_PROJECT_LOCATION + "-"+ REDCAP_RECORD_ID_2;
    public static final String REDCAP_SUBJECT_ID_FIELD = "subject_id";

    @Before
    public void init() throws IOException, TokenException {
        redCapClient = Mockito.mock(RedCapClient.class);
        redCapIntegrator = new RedCapIntegator(redCapClient);
        testAttributes.put(REDCAP_ATTRIBUTE_1,REDCAP_ATTRIBUTE_1_VAL);
        testAttributes.put(REDCAP_ATTRIBUTE_2,REDCAP_ATTRIBUTE_2_VAL);
        testAttributes.put(HUMAN_READABLE_ID_KEY,HUMAN_READABLE_ID);
        attributeKeys = redCapInfo.getAttributeFieldNames();

        updateProjectAttributes();
    }

    @Test
    public void getInputTest() {
        Integrator integrator = new Integrator(trigger, mpClient);

        assertEquals(integrator.handleDataEntryTrigger(), false);
    }

    @Test
    public void updateAttributesInMpTest() throws IOException, URISyntaxException {
        Mockito.when(redCapClient.fetchFormDataForId(attributeKeys, REDCAP_RECORD_ID_2)).thenReturn(testAttributes);
        Map<String, String> attributes = redCapIntegrator.pullRecordAttributes(attributeKeys, REDCAP_RECORD_ID_2);

        assertEquals(testAttributes, attributes);

        Subject subject = mpIntegrator.performSubjectUpdateOnMp(redCapInfo.getUrl(), redCapInfo.getProjectId(), REDCAP_RECORD_ID_2, attributes, "");

        assertEquals(attributes, subject.getAttributes());

        assertEquals(HUMAN_READABLE_ID, subject.getHumanReadableIdentifier());
    }

    @Test
    public void updateAttributesInMpWhenSubjectExistsTest() throws IOException, URISyntaxException {
        testAttributes.put(REDCAP_ATTRIBUTE_1, REDCAP_ATTRIBUTE_1_VAL_2);
        Mockito.when(redCapClient.fetchFormDataForId(attributeKeys, REDCAP_RECORD_ID_2)).thenReturn(testAttributes);
        Map<String, String> attributes = redCapIntegrator.pullRecordAttributes(attributeKeys, REDCAP_RECORD_ID_2);

        String existingSubjectId = mpClient.getSubject(redCapInfo.getUrl(), redCapInfo.getProjectId(), REDCAP_RECORD_ID_2).getSubjectId();

        Subject subject = mpIntegrator.performSubjectUpdateOnMp(redCapInfo.getUrl(), redCapInfo.getProjectId(), REDCAP_RECORD_ID_2, attributes,existingSubjectId);

        assertEquals(attributes, subject.getAttributes());
        assertEquals(REDCAP_ATTRIBUTE_1_VAL_2, subject.getAttributes().get(REDCAP_ATTRIBUTE_1));

        assertEquals(HUMAN_READABLE_ID, subject.getHumanReadableIdentifier());
    }

    @Test
    public void createSubjectWhenSubjectExistsInMpTest() throws IOException, URISyntaxException {
        Subject subject = mpIntegrator.performSubjectUpdateOnMp(redCapInfo.getUrl(), redCapInfo.getProjectId(), REDCAP_RECORD_ID_2, new HashMap<>(),"");

        // This should fail since MP project exists but REDCap project does not exists/was deleted
        assertEquals(Subject.SubjectOperationStatus.FAILED, subject.getOprationStatus());
    }
}

File: src/integration-test/kotlin/org/radarbase/redcap/integration/IntegratorTest.kt
package org.radarbase.redcap.integration

import com.nhaarman.mockitokotlin2.doReturn
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import org.junit.Assert
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Before
import org.junit.FixMethodOrder
import org.junit.Test
import org.junit.runners.MethodSorters
import org.radarbase.redcap.config.RedCapInfo
import org.radarbase.redcap.config.RedCapManager.getInfo
import org.radarbase.redcap.managementportal.Subject
import org.radarbase.redcap.util.IntegrationUtils
import org.radarbase.redcap.util.IntegrationUtils.LOGGER
import org.radarbase.redcap.util.IntegrationUtils.MP_PROJECT_ID
import org.radarbase.redcap.util.IntegrationUtils.MP_PROJECT_LOCATION
import org.radarbase.redcap.util.IntegrationUtils.REDCAP_RECORD_ID_1
import org.radarbase.redcap.util.IntegrationUtils.REDCAP_RECORD_ID_2
import org.radarbase.redcap.util.IntegrationUtils.TRIGGER_BODY
import org.radarbase.redcap.util.IntegrationUtils.WORK_PACKAGE
import org.radarbase.redcap.util.IntegrationUtils.mpClient
import org.radarbase.redcap.util.IntegrationUtils.updateProjectAttributes
import org.radarbase.redcap.util.RedCapClient
import org.radarbase.redcap.util.RedCapTrigger
import java.io.IOException
import java.net.URISyntaxException
import java.net.URL
import java.util.*

@FixMethodOrder(MethodSorters.NAME_ASCENDING)
class IntegratorTest {
    private val mpIntegrator =
        MpIntegrator(mpClient)
    private val trigger = RedCapTrigger(TRIGGER_BODY)
    private val redCapInfo: RedCapInfo = getInfo(trigger)

    private val testAttributes: MutableMap<String, String> = mutableMapOf(
        Pair(
            REDCAP_ATTRIBUTE_1,
            REDCAP_ATTRIBUTE_1_VAL
        ),
        Pair(
            REDCAP_ATTRIBUTE_2,
            REDCAP_ATTRIBUTE_2_VAL
        )
    )

    private val attributeKeys: List<String> = listOf(
        REDCAP_ATTRIBUTE_1,
        REDCAP_ATTRIBUTE_2,
        IntegrationData.SUBJECT_ID_LABEL
    )

    private val redCapClient: RedCapClient = mock {
        on { fetchFormDataForId(attributeKeys, REDCAP_RECORD_ID_1) } doReturn testAttributes
    }
    private val redCapIntegrator: RedCapIntegator =
        RedCapIntegator(redCapClient)

    @Before
    fun init() = updateProjectAttributes()

    @Test
    fun inputTest() {

        // We can check if corresponding subject exists in MP too.
        val subject: Subject? = mpClient.getSubject(
            URL(IntegrationUtils.REDCAP_URL),
            IntegrationUtils.REDCAP_PROJECT_ID,
            REDCAP_RECORD_ID_2
        )

        LOGGER.info("Subject: {}", subject)

        val attr =
            if (subject != null) {
                testAttributes.toMutableMap().apply {
                    this[IntegrationData.SUBJECT_ID_LABEL] = subject.subjectId
                }
            } else testAttributes

        LOGGER.info("Attributes: {}", attr.entries.toTypedArray())

        whenever(
            redCapClient.fetchFormDataForId(
                attributeKeys,
                REDCAP_RECORD_ID_2
            )
        ).thenReturn(
            attr
        )

        val integrator = Integrator(
            trigger = trigger,
            mpClient = mpClient,
            redCapInfo = redCapInfo,
            redCapIntegrator = redCapIntegrator,
            mpIntegrator = mpIntegrator
        )

        // The result will be false since there is no actual redcap instance and hence the form
        // cannot be updated. But even then, no exception should be thrown.
        val result = integrator.handleDataEntryTrigger()
        assertFalse(result)

        Assert.assertNotNull(subject)
        assertEquals(
            Integer.valueOf(REDCAP_RECORD_ID_2),
            subject?.externalId
        )
        assertEquals(
            "$WORK_PACKAGE-$MP_PROJECT_ID-$MP_PROJECT_LOCATION-" +
                    "$REDCAP_RECORD_ID_2",
            subject?.humanReadableIdentifier
        )
        assertEquals("radar", subject?.project?.projectName)
    }

    @Test
    @Throws(IOException::class, URISyntaxException::class)
    fun updateAttributesInMpTest() {
        val attributes: MutableMap<String, String> =
            redCapIntegrator.pullFieldsFromRedcap(attributeKeys, REDCAP_RECORD_ID_1)
        assertEquals(attributes, testAttributes)
        val existingSubjectId: String =
            mpClient.getSubject(redCapInfo.url, redCapInfo.projectId, REDCAP_RECORD_ID_1)!!
                .subjectId
        val subject = mpIntegrator.performSubjectUpdateOnMp(
            redCapInfo.url,
            redCapInfo.projectId,
            REDCAP_RECORD_ID_1,
            attributes,
            existingSubjectId
        )
        assertEquals(subject.attributes, attributes)
        assertEquals(subject.humanReadableIdentifier,
            HUMAN_READABLE_ID
        )
    }

    @Test
    @Throws(IOException::class, URISyntaxException::class)
    fun updateAttributesInMpWhenSubjectExistsTest() {
        testAttributes[REDCAP_ATTRIBUTE_1] =
            REDCAP_ATTRIBUTE_1_VAL_2
        whenever(redCapClient.fetchFormDataForId(attributeKeys, REDCAP_RECORD_ID_1)).thenReturn(
            testAttributes
        )
        val attributes: MutableMap<String, String> =
            redCapIntegrator.pullFieldsFromRedcap(attributeKeys, REDCAP_RECORD_ID_1)

        val existingSubjectId: String =
            mpClient.getSubject(redCapInfo.url, redCapInfo.projectId, REDCAP_RECORD_ID_1)!!
                .subjectId
        val subject = mpIntegrator.performSubjectUpdateOnMp(
            redCapInfo.url,
            redCapInfo.projectId,
            REDCAP_RECORD_ID_1,
            attributes,
            existingSubjectId
        )
        assertEquals(subject.attributes, attributes)
        assertEquals(subject.attributes[REDCAP_ATTRIBUTE_1],
            REDCAP_ATTRIBUTE_1_VAL_2
        )
        assertEquals(subject.humanReadableIdentifier,
            HUMAN_READABLE_ID
        )
    }

    @Test
    @Throws(IOException::class, URISyntaxException::class)
    fun createSubjectWhenSubjectExistsInMpTest() {
        val subject = mpIntegrator.performSubjectUpdateOnMp(
            redCapInfo.url,
            redCapInfo.projectId,
            REDCAP_RECORD_ID_1,
            HashMap(),
            ""
        )
        // This should fail since MP subject exists but REDCap subject does not exists/was deleted
        assertEquals(Subject.SubjectOperationStatus.FAILED, subject.operationStatus)
    }

    companion object {
        private const val REDCAP_ATTRIBUTE_1 = "treatment_arm"
        private const val REDCAP_ATTRIBUTE_1_VAL = "arm1"
        private const val REDCAP_ATTRIBUTE_1_VAL_2 = "arm2"
        private const val REDCAP_ATTRIBUTE_2 = "another_attribute"
        private const val REDCAP_ATTRIBUTE_2_VAL = "test"
        private const val HUMAN_READABLE_ID_KEY = "Human-readable-identifier"
        private const val HUMAN_READABLE_ID: String =
            "$WORK_PACKAGE-$MP_PROJECT_ID-$MP_PROJECT_LOCATION-$REDCAP_RECORD_ID_1"
    }

}